### ISSUE TITLE
Reduced words of ReflectionGroup elements are not well-defined

### ISSUE THREAD

**Issue by tscrim** [sagemath/sage#34912]
<div id="comment:0"></div>

We cannot use it for hashing as I found the following:

```
sage: G4 = ReflectionGroup(4)
sage: elt = G4[22]
sage: elt
(1,12)(2,24)(3,19)(4,22)(5,17)(6,20)(7,23)(8,9)(10,21)(11,13)(14,18)(15,16)
sage: y = (elt * G4.gen(1)) * G4.gen(1) * G4.gen(1)
sage: elt == y
True
sage: hash(elt) == hash(y)
False
sage: elt._reduced_word
[0, 0, 1, 0, 0, 1]
sage: y._reduced_word
[1, 0, 0, 1, 0, 0]
```
Either we figure out how to make this consistent or we need to change the hashing.

CC:  @stumpc5

Component: **group theory**

Keywords: **gap3, complex reflection group**

Author: **Travis Scrimshaw, Christian Stump**

Branch/Commit: **[u/tscrim/hash_reflection_groups-34912](https://github.com/sagemath/sagetrac-mirror/tree/u/tscrim/hash_reflection_groups-34912) @ [`14d6f61`](https://github.com/sagemath/sagetrac-mirror/commit/14d6f6191190904f92728e00989de7ec1b7ff180)**

Reviewer: **Christian Stump, Travis Scrimshaw**

_Issue created by migration from https://trac.sagemath.org/ticket/34912_

**Comment by tscrim** [comment]
Description changed:
``````diff
--- 
+++ 
@@ -1,6 +1,7 @@
 We cannot use it for hashing as I found the following:
 
 ```
+sage: G4 = ReflectionGroup(4)
 sage: elt = G4[22]
 sage: elt(1,12)(2,24)(3,19)(4,22)(5,17)(6,20)(7,23)(8,9)(10,21)(11,13)(14,18)(15,16)
 sage: y = (elt * G4.gen(1)) * G4.gen(1) * G4.gen(1)
``````

**Comment by tscrim** [comment]
Description changed:
``````diff
--- 
+++ 
@@ -3,7 +3,8 @@
 ```
 sage: G4 = ReflectionGroup(4)
 sage: elt = G4[22]
-sage: elt(1,12)(2,24)(3,19)(4,22)(5,17)(6,20)(7,23)(8,9)(10,21)(11,13)(14,18)(15,16)
+sage: elt
+(1,12)(2,24)(3,19)(4,22)(5,17)(6,20)(7,23)(8,9)(10,21)(11,13)(14,18)(15,16)
 sage: y = (elt * G4.gen(1)) * G4.gen(1) * G4.gen(1)
 sage: elt == y
 True
``````

**Comment by stumpc5** [comment]
<div id="comment:3" align="right">comment:3</div>

I do not remember why we chose to use that gap3 method to compute reduced words rather than taking the lex-minimal one in the simples (which is also very fast and much cleaner to implement).

We could try to move to the lex-min one after checking the timings.

**Comment by tscrim** [comment]
<div id="comment:4" align="right">comment:4</div>

I think this demonstrates that we need to go with the lex-min, unless we change the hash function. We can keep both, with an option in `reduced_word()` to choose between the two, but it probably would be best just have the lex-min only (not just the default or only for the hashing).

**Comment by stumpc5** [comment]
Branch: **[u/stumpc5/reduced_words_of_reflectiongroup_elements_are_not_well_defined](https://github.com/sagemath/sagetrac-mirror/tree/u/stumpc5/reduced_words_of_reflectiongroup_elements_are_not_well_defined)**

**Comment by stumpc5** [comment]
<div id="comment:6" align="right">comment:6</div>

@tscrim: it turned out that we cannot use the reduced word for hashing since recomputing it takes too long. Please play around with `test_hash` and use move it to `__hash__` when you are satisfied.

**Comment by stumpc5** [comment]
Commit: **[`498e8db`](https://github.com/sagemath/sagetrac-mirror/commit/498e8dbd59906fcd80ff10d5f3f66b0168cd5abc)**

**Comment by tscrim** [comment]
<div id="comment:7" align="right">comment:7</div>

Are you saying it is too slow to just cache the lex-min reduced word or that it is too slow to have both (potentially computing the lex-min every time `__hash__` is called)?

I was suggesting to have both functions that can create both ways. Although I meant to say that the default (and what we cache) should be the lex-min.

If computing the lex-min is too slow, then let's just use the hash of the permutation representation with

```python
def __hash__(self):
    return hash(self._parent) | super().__hash__()
```

**Comment by stumpc5** [comment]
<div id="comment:8" align="right">comment:8</div>

No, we do not have any good way of computing reduced words for complex non-real reflection groups. That means that we currently have

*  an interator that caches reduced words along the way (`for w in W: print(w.reduced_word())`). That is reasonably fast, but not stable, meaning that we cannot assume to get the same reduced word along iterations (though it usually will be the same).

* a method `reduced_word` that recomputes it using gap3 or gap4 depending on a permutation group with generators. That is slow, but can be done if really needed.

Neither of the two is good for hashing, one is not stable and one is too slow. But the permutation representation is really fast and suitable for hashing.

**Comment by stumpc5** [comment]
<div id="comment:9" align="right">comment:9</div>

Please compare `super().__hash__()` with my hash method and just use the faster one. Yours is probably the right one to use.

**Comment by tscrim** [comment]
<div id="comment:10" align="right">comment:10</div>

Okay, thank you. I will do that tomorrow.

**Comment by tscrim** [comment]
<div id="comment:11" align="right">comment:11</div>

Using the hash from `super()` is faster:

```
sage: G20 = ReflectionGroup(20)
sage: G20.cardinality()
360
sage: %timeit [hash(x) for x in G20]
1.2 ms ± 16 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
sage: %timeit [x.test_hash() for x in G20]
1.45 ms ± 21.1 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

sage: W = ReflectionGroup((4,2,5))
sage: W.cardinality()
61440
sage: %timeit [x.__hash__() for x in W]
36.2 ms ± 1.06 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
sage: %timeit [x.test_hash() for x in W]
91.3 ms ± 776 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

**Comment by tscrim** [comment]
Changed branch from **[u/stumpc5/reduced_words_of_reflectiongroup_elements_are_not_well_defined](https://github.com/sagemath/sagetrac-mirror/tree/u/stumpc5/reduced_words_of_reflectiongroup_elements_are_not_well_defined)** to **[u/tscrim/hash_reflection_groups-34912](https://github.com/sagemath/sagetrac-mirror/tree/u/tscrim/hash_reflection_groups-34912)**

**Comment by tscrim** [comment]
Author: **Travis Scrimshaw, Christian Stump**

**Comment by tscrim** [comment]
<div id="comment:12" align="right">comment:12</div>

I have changed it and updated some doctests. I also fixed a few others that I saw along the way with gap3 not being regularly tested against.

#30450 no longer seems to occur (both before and after this ticket).

The `gap3` interface file tests all seem to be quite broken, but that is unrelated to this ticket.

---
New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/518ecbfb3a62dd809f090b15fe6158344f6b3077"><code>518ecbf</code></a></td><td><code>Merge branch 'u/stumpc5/reduced_words_of_reflectiongroup_elements_are_not_well_defined' of https://github.com/sagemath/sagetrac-mirror into u/tscrim/hash_reflection_groups-34912</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/3a1a03ae2cc38d75691418bfda23a703e002139c"><code>3a1a03a</code></a></td><td><code>Updating `__hash__` and updating gap3 doctests to pass.</code></td></tr></table>

**Comment by tscrim** [comment]
Reviewer: **Christian Stump, Travis Scrimshaw**

**Comment by tscrim** [comment]
Changed commit from **[`498e8db`](https://github.com/sagemath/sagetrac-mirror/commit/498e8dbd59906fcd80ff10d5f3f66b0168cd5abc)** to **[`3a1a03a`](https://github.com/sagemath/sagetrac-mirror/commit/3a1a03ae2cc38d75691418bfda23a703e002139c)**

**Comment by stumpc5** [comment]
<div id="comment:13" align="right">comment:13</div>

I would only use `iterator_tracking_words` if actually needed. In the other situations, we should keep `for w in W:` to check that the iterator (inherited from permutation groups) also works correctly.

Otherwise, the ticket looks good to go!

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`3a1a03a`](https://github.com/sagemath/sagetrac-mirror/commit/3a1a03ae2cc38d75691418bfda23a703e002139c)** to **[`14d6f61`](https://github.com/sagemath/sagetrac-mirror/commit/14d6f6191190904f92728e00989de7ec1b7ff180)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:14"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/14d6f6191190904f92728e00989de7ec1b7ff180"><code>14d6f61</code></a></td><td><code>Adding some systematic doctests and one detail.</code></td></tr></table>

**Comment by tscrim** [comment]
<div id="comment:15" align="right">comment:15</div>

Thank you. I didn't want to update those doctests and outputs, so instead I just added some equivalent systematic checks that made sure the results were equivalent.

**Comment by tscrim** [comment]
<div id="comment:17" align="right">comment:17</div>

Thank you!

**Comment by mkoeppe** [comment]
Merged in 10.0.beta0
